src/analysis/normed_space/box_subadditive.lean

```diff
@@ -0,0 +1,541 @@
+
+protected lemma neg [add_comm_group G] {f : (ι → α) → (ι → α) → G} (hf : box_additive_on f s) :
+  box_additive_on (-f) s :=
+λ l u h m hm i, by simp [← hf h hm i, add_comm]
+
+protected lemma sub [add_comm_group G] {f g : (ι → α) → (ι → α) → G}
+  (hf : box_additive_on f s) (hg : box_additive_on g s) :
+  box_additive_on (f - g) s :=
+by simp only [sub_eq_add_neg, hf.add hg.neg]
+
+protected lemma prod [fintype ι] {R} [comm_semiring R] (f : α → α → R)
+  (hf : ∀ ⦃x y z⦄, x ≤ y → y ≤ z → f x y + f y z = f x z) :
+  box_additive_on (λ x y, ∏ i : ι, f (x i) (y i)) s :=
+begin
+  intros l u h m hm i,
+  have := function.apply_update (λ j, f (l j)) u i (m i),
+  have := function.apply_update (λ j y, f y (u j)) l i (m i),
+  simp only at *,
+  simp only [*, prod_update_of_mem, mem_univ, ← add_mul],
+  rw [← prod_mul_prod_compl {i}, prod_singleton, compl_eq_univ_sdiff, hf (hm.1 i) (hm.2 i)]
+end
+
+protected lemma box_subadditive_on [ordered_add_comm_monoid M] {f : (ι → α) → (ι → α) → M}
+  (hf : box_additive_on f s) : box_subadditive_on f s :=
+λ l u h m hm i, (hf h hm i).ge
+
+protected lemma box_supadditive_on [ordered_add_comm_monoid M] {f : (ι → α) → (ι → α) → M}
+  (hf : box_additive_on f s) : box_supadditive_on f s :=
+λ l u h m hm i, (hf h hm i).le
+
+lemma norm_subadditive_on {E : Type*} [normed_group E] {f : (ι → α) → (ι → α) → E}
+  (hf : box_additive_on f s) : box_subadditive_on (λ x y, ∥f x y∥) s :=
+λ l u h m hm i, by simp only [← hf h hm i, norm_add_le]
+
+end box_additive_on
+
+namespace box_supadditive_on
+
+variables [decidable_eq ι] [preorder α] [ordered_add_comm_monoid M] {s : set (ι → α)}
+  {l m u : ι → α} {f : (ι → α) → (ι → α) → M}
+
+protected lemma order_dual (hf : box_supadditive_on f s) :
+  @box_subadditive_on ι α (order_dual M) _ _ _ f s :=
+hf
+
+protected lemma mono (h : box_supadditive_on f s) {t} (ht : t ⊆ s) : box_supadditive_on f t :=
+λ l u hsub, h (set.subset.trans hsub ht)
+
+lemma sum_finset_subboxes_le (h : box_supadditive_on f s) (hsub : Icc l u ⊆ s) (hm : m ∈ Icc l u)
+  (t : finset ι) :
+  ∑ t' in t.powerset, f (t'.piecewise m l) ((t \ t').piecewise m u) ≤ f l u :=
+h.order_dual.le_sum_finset_subboxes hsub hm t
+
+variables [fintype ι]
+
+/-- Take a rectangular box `[lo, hi]` in `ι → α` and a point `mid ∈ [lo, hi]`. The hyperplanes `x i
+= mid i` split the box `[lo, hi]` into `2^n` subboxes, where `n = card ι`.  If `f` is supadditive on
+subboxes, then its value on `[lo, hi]` is greater than or equal to the sum of its values on these
+`2^n` subboxes. -/
+lemma sum_subboxes_le (h : box_supadditive_on f s) (hsub : Icc l u ⊆ s) (hm : m ∈ Icc l u) :
+  ∑ t : finset ι, f (t.piecewise m l) (t.piecewise u m) ≤ f l u :=
+h.order_dual.le_sum_subboxes hsub hm
+
+end box_supadditive_on
+
+section coe
+
+variables {N : Type*} [decidable_eq ι] [preorder α] {s : set (ι → α)}
+
+lemma box_subsupadditive_coe_helper [add_monoid M] [add_monoid N] {c : M → N} (rM : M → M → Prop)
```

Comment:
This lemma is unreadable, and needs a docstring (if only to say that this is a triviality used below for rewriting).